### ISSUE TITLE
ASP.NET Project containing any server-side Blazor components is compiled twice

### ISSUE THREAD

**Issue by willdean** [dotnet/aspnetcore#17742]
### Describe the bug
I don't actually know if this is a bug or if it's a temporary impediment or what, but when an ASP.NET Core project which contains a Razor Component is compiled, the entire source appears to be compiled twice, which seems a bit sad with respect to perf.

### To Reproduce
* Create default server-side Blazor project
* Add something to startup.cs which causes a compiler warning - e.g. 
```
       return;
       Console.WriteLine("Hello"); 
```
at the end of `ConfigureServices()`
* Build the project
* Observe that the compiler output contains:
```
1>Startup.cs(34,13,34,20): warning CS0162: Unreachable code detected
1>Startup.cs(34,13,34,20): warning CS0162: Unreachable code detected
```

We've observed this on a much larger ASP.net Core project to which we're adding Razor components - it's clear from warning messages that the entire source code, (CS, CSHTML, etc),  is being compiled twice.

Is this expected behaviour for a server-side application?    Are there plans to optimise it in the future?
Apologies if this is discussed elsewhere already, but I did look and I couldn't find it.

### Further technical details
- ASP.NET Core version 3.1.100
- VS2019 16.4

**Comment by javiercn** [comment]
@willdean thanks for contacting us.

This is expected as components need to be able to reference each other so in order to be able to successfully provide tooling support Razor does a two phase compilation where it first compiles a "declaration" of the components and then produces the final compilation with all the appropriate components.

This is similar to what other technologies like XAML do and is very common. We haven't seen any perf issue so far, but if you have data of a real use case where this presents an actual issue (it increases the build time in minutes and represents the majority of the build time), we would take a look at the data. So far, we haven't found any issue.

With regards to showing the errors twice. @pranavkm @rynowak anything we can/care to do in this space?

**Comment by willdean** [comment]
@javiercn Thanks for the reply - I can't say that I've previously noticed WPF projects building twice like this (if that's what XAML implied), but I will look more closely!  
I had noticed that certain 'Blazor-specific' errors only happen on the second compile, whereas warnings in unrelated code appear twice.
I can't claim this is a real-world ("minutes") problem, but I have infinite supplies of impatience, and current Blazor development does force one around and around a complete build/restart loop a lot, so any evidence that time might be being wasted is seized on.   
If nothing else, suppressing the duplicate errors would stop people like me complaining about the (possibly hypothetical) waste of time.

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us. We will look into this after we're dong with the upcoming Blazor WASM release.

**Comment by redowl3** [comment]
@javiercn does the location of the components make any difference - if for example you have a number of folders for each type of component (e.g. membership, products)?

**Comment by javiercn** [comment]
@willdean No.

**Comment by rynowak** [comment]
We're looking at making some improvements to how this works in 5.0 as part of the Roslyn source generators work.

**Comment by pranavkm** [comment]
Moving this to backlog. This is currently blocked on source generators effort.

**Comment by adrianwright109** [comment]
@pranavkm is this issue a duplicate of #26902?
